[Concurrency] Provide a Swift interface for custom main and global executors.

al45tair opened the pull request:
Reorganise the Concurrency code so that it's possible to completely implement executors (both main and global) in Swift.

Provide API to choose the desired executors for your application.

Also make `Task.Sleep` wait using the current executor, not the global executor, and expose APIs on `Clock` to allow for conversion between time bases.

rdar://141348916

al45tair committed:
[Concurrency] Swift interface for custom main and global executors.

Reorganise the Concurrency code so that it's possible to completely
implement executors (both main and global) in Swift.

Provide API to choose the desired executors for your application.

Also make `Task.Sleep` wait using the current executor, not the global
executor, and expose APIs on `Clock` to allow for conversion between
time bases.

rdar://141348916

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
(This is not yet complete; there's a problem I need to resolve relating to setting the main executor.)

ktoso reviewed (commented):
Just a few minor comments so far; overall this is looking great :)

al45tair committed:
[Concurrency] More work on the custom executor implementation.

Added an `-executor-factory` argument to the compiler to let you safely
specify the executors you wish to use (by naming a type that returns
them).

Also added some tests of the new functionality.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency] Add clock traits.

Remove the hacky support for mapping clocks to a Dispatch clock ID,
in favour of clocks publishing traits and having the Dispatch
executor select the clock on the basis of those traits.

al45tair committed:
[Concurrency] Fix potential ABI breakages.

Fix a couple of potential ABI breaks.  Also add the new functions and types
to the baseline lists.

rdar://141348916

al45tair committed:
[Concurrency] Rename ExecutorJob back to PartialAsyncTask.

The rename appears to have had some kind of potentially ABI breaking
consequence.

rdar://141348916

al45tair committed:
[Concurrency] Remove spurious printf().

Remove a printf() I'd added for debugging.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency][Embedded] Remove MainActor/MainExecutor everywhere.

Embedded Swift doesn't have MainActor, so remove it.

rdar://141348916

al45tair committed:
[Concurrency] Don't use blocks.

We can't use blocks, because Swift doesn't support them on Linux or
Windows.  Instead, use a C function pointer, and box up the handler.

rdar://141348916

al45tair commented on stdlib/public/Concurrency/CMakeLists.txt:
```diff
@@ -134,10 +124,10 @@ set(SWIFT_RUNTIME_CONCURRENCY_SWIFT_SOURCES
   AsyncThrowingFlatMapSequence.swift
   AsyncThrowingMapSequence.swift
   AsyncThrowingPrefixWhileSequence.swift
+  ExecutorJob.swift
```
Sadly I've had to row back on this for now because of a bug in the ABI checker.

al45tair commented on stdlib/public/Concurrency/DispatchExecutor.swift:
```diff
@@ -0,0 +1,226 @@
+// .. Task Executor ............................................................
+
+@available(SwiftStdlib 6.2, *)
+public class DispatchTaskExecutor: TaskExecutor, @unchecked Sendable {
```
Would we not call that `DispatchExecutor`? (I have a protocol of that name, so I'll rename that to `DispatchExecutorProtocol`.)

al45tair replied:
Or, I guess, maybe `DispatchQueueExecutor`? I don't think we'd call it `DispatchTaskExecutor`, would we?

al45tair replied:
Hmmm. Or maybe we would. I've renamed it to `DispatchGlobalTaskExecutor`, and also renamed the protocol.

al45tair committed:
[Concurrency] Fixes from initial review.

Rename `DispatchTaskExecutor` to `DispatchGlobalTaskExecutor` as we
may want to use the former for an executor that runs things on an
arbitrary Dispatch queue.

Rename `DispatchExecutor` to `DispatchExecutorProtocol`; again, we
might want the name for something else.

Add `@Sendable` attribute to `registerEvent`.

Fix missing `extern "C" SWIFT_CC(swift)` on `_swift_exit` (merge
error).

Remove stray whitespace from `CMakeLists.txt`

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency] Mark expressions as `unsafe`.

Also fix a missing availability annotation.

rdar://141348916

al45tair committed:
[Concurrency] Update header file.

This needed an update to make things build on Linux and Windows.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency] Update following pitch comments.

Remove `supportsScheduling` in favour of a type-based approach.

Update the storage for `ClockTraits` to `UInt32`.

Adjust ordering of executors for `currentExecutor`.

rdar://141348916

al45tair committed:
[Concurrency] Remove EventableExecutor, alter asSchedulable.

EventableExecutor is being removed, for now, but hopefully will
return in some form in the future.

The `asSchedulable` implementation needs to change for reasons of
ABI stability.

rdar://141348916

al45tair committed:
[Concurrency] Make `currentExecutor` return a non-optional.

If we don't have any other executor to return, return the default
executor.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency][Tests] Fix baseline-asserts.

Apparently I hadn't included the change for `currentExecutor` being a
non-`Optional`.

rdar://141348916

al45tair committed:
[Concurrency][Tests] Linux doesn't have DispatchTime.advanced(by:).

We need instead to use the `+` operator.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency][Tests] Remove spurious line in baseline-asserts.

I think this might have been a merge error at some point while rebasing.

rdar://141348916

al45tair committed:
[Concurrency] Add a missing `public`.

The default implementation of `traits` on `Clock` needs to be `public`.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency][Linux] Fix WASI build.

This doesn't actually add proper WASI support - it just fixes the
build so that we're no worse off than we were.

rdar://141348916

al45tair committed:
[Concurrency][Tests] Add missing `traits` symbol to baseline-asserts.

Having made it public, we need to add it to baseline-asserts.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair commented:
SE proposal is here: https://github.com/swiftlang/swift-evolution/pull/2654

al45tair commented:
WASI concurrency tests failed, which is not a great surprise, as I've only just discovered that those were a thing.

al45tair committed:
[Concurrency] Add CooperativeExecutor, use it.

Also tweak the sleep implementations to let the hooks run if there
isn't a `SchedulableExecutor` (for hooked mode).

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency] Fix availability.

Fix up availability after the non-Darwin changes.  Also update the
ABI baseline.

Fix a Win32 typo.

rdar://141348916

al45tair committed:
[Concurrency] Another missing availability annotation.

We need to annotate the `timestamp()` function for watchOS (apparently
not for macOS for some reason).

rdar://141348916

al45tair committed:
[Concurrency] Fix a typo and use `var` not `let`.

Fix a typo that caused us to not include the correct code for platforms
that use "singlethreaded" concurrency.

Also use `var` not `let` for the computed property in `PlatformExecutorFactory`.

rdar://141348916

al45tair committed:
[Concurrency][Tests] Require `libdispatch` in the sleep test.

We need to explicitly require `libdispatch` here, otherwise the
Windows build will fail.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency][32-bit] Store timestamp out-of-line on 32-bit.

CooperativeExecutor needs to store its timestamp out-of-line for
32-bit architectures (like WASI).

rdar://141348916

al45tair committed:
[Concurrency] Fall back to allocating from the heap, fix a test.

`CooperativeExecutor` should fall back to heap allocations if the task
allocator isn't available when it's enqueuing delayed jobs.

Also, remove the reference to `DispatchGlobalExecutor` from the
custom executor test.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency] Fix some warnings, use typed throws.

Fix a couple of `unsafe` warnings.

Also update `withUnsafeExecutorPrivateData` to use typed throws instead of
`rethrow`.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test Linux platform

al45tair commented:
OK. Everything is now passing except for the minimalstdlib configuration.

al45tair committed:
[Concurrency] Disable various things for task-to-thread model.

When in task-to-thread model concurrency mode, there is no `MainActor`
and we cannot use `ExecutorJob`, so disable various things.

rdar://141348916

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -35,6 +35,134 @@ public protocol Executor: AnyObject, Sendable {
   #endif // !SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY
+
+  #if !$Embedded && !SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY
+  // The functions below could have been added to a separate protocol,
+  // but doing that would then mean doing an `as?` cast in e.g.
+  // enqueueOnGlobalExecutor (in ExecutorBridge.swift), which is
+  // undesirable from a performance perspective.
```
This comment is a hang-over from an older version and should be removed.

al45tair commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -319,6 +461,192 @@ extension SerialExecutor {
+  ///   2. The preferred executor for the currently executing `Task`, or
+  ///   3. The task executor for the current thread
+  ///
+  ///  If none of these exist, this property will be `nil`.
```
This last line is no longer true; it will never be `nil`.

ktoso replied:
I think that's reasonable hm... Let's see if SE review ends up with some more requests.

If not then we should probably clarify that a non-concurrency thread would end up returning the default executor - which seems fine

al45tair commented on stdlib/public/Concurrency/ExecutorImpl.swift:
```diff
@@ -0,0 +1,100 @@
+
+@available(SwiftStdlib 6.2, *)
+@_silgen_name("swift_task_donateThreadToGlobalExecutorUntilImpl")
+internal func dontateToGlobalExecutor(
```
Typo. Should be `donateToGlobalExecutor`.

al45tair commented on test/Concurrency/Runtime/clocks.swift:
```diff
@@ -0,0 +1,276 @@
+
+  // These are a bit of a lie, since this clock is weird and doesn't
+  // actually tell the time; for the purposes of this test, we pretend
+  // that the tocks are 20ms.
```
We actually pretend they're 10ms.

al45tair reviewed (commented):
Spotted a couple of mistakes in comments.

ktoso commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -319,6 +461,192 @@ extension SerialExecutor {
+  @available(SwiftStdlib 6.2, *)
+  public func isSameExclusiveExecutionContext(other: Self) -> Bool {
+    return self == other
+  }
```
Are we changing in any way what is expected to be invoked when? If we're moving away from the using the returned executor then we may need to update the diagram:  https://github.com/swiftlang/swift-evolution/pull/2716/files#diff-a8714de4afecd9377aa16f6cc8ed9297e84a9c7b109feb6fe18246742c9e02b9

ktoso commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -319,6 +461,192 @@ extension SerialExecutor {
+  /// soon as is practicable, stop the innermost `run()` invocation and make
+  /// that `run()` invocation return.
+  func stop()
+}
```
good docs on the type overall

ktoso commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -370,17 +698,33 @@ public struct UnownedSerialExecutor: Sendable {
+    } else {
+      unsafe self.executor = Builtin.buildOrdinarySerialExecutorRef(executor)
+    }
+  }
```
Aha, got it -- that's the bit I was wondering about up-thread; This is quite nice, I think it's ok to use this new style of opting in, very nice! 👍

ktoso replied:
Does this mean though that we should deprecate the other old initializers?

ktoso commented on stdlib/public/Concurrency/ExecutorBridge.cpp:
```diff
@@ -0,0 +1,134 @@
+      0, "Incorrect actor executor assumption; expected '%.*s' executor.\n",
+      (int)typeName.length, typeName.data);
+  }
+}
```
looks good, thanks for moving these over to ExecutorBridge now

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -370,6 +446,82 @@ extension ExecutorJob {
+  ///
+  /// N.B. Because this allocator is stack disciplined, explicitly
+  /// deallocating memory will also deallocate all memory allocated
+  /// after the block being deallocated.
```
I think I had this question before already: that's now how we behave today, if you deallocate "in wrong order" we crash; are we really changing that? That was good for detecting we messed up somehow, while deallocating other things like this may hide bugs in stack discipline...?

ktoso commented on stdlib/public/Concurrency/PlatformExecutorLinux.swift:
```diff
@@ -0,0 +1,25 @@
+    = DispatchGlobalTaskExecutor()
+}
+
+#endif // os(Linux)
```
I really like how those Platform...swift look now 👍

ktoso commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -377,11 +377,17 @@ static SerialExecutorRef executorForEnqueuedJob(Job *job) {
+  //if (auto identity = reinterpret_cast<HeapObject *>(jobQueue)) {
+  //  return SerialExecutorRef::forOrdinary(
+  //      identity, _swift_task_getDispatchQueueSerialExecutorWitnessTable());
+  //}
```
remove?

ktoso commented on test/Concurrency/Runtime/custom_main_executor.swift:
```diff
@@ -0,0 +1,92 @@
+// RUN: %target-run-simple-swift(-Xfrontend -disable-availability-checking -g %import-libdispatch -parse-as-library -executor-factory SimpleExecutorFactory) | %FileCheck %s
```
Do we need a test specifying the executor factory in another module?

ktoso reviewed (commented):
LGTM! Just small things here and there, feel free to merge and let's follow up on what makes sense to address 👍

al45tair commented:
> LGTM! Just small things here and there, feel free to merge and let's follow up on what makes sense to address 👍

Sounds good. Will merge and then make a new PR to address the various things noted on this one.
